Supporting _include:iterate and _revinclude:iterate

limorl opened the pull request:
## Description
The :iterate modifier (:recurse in STU3)  allows to query iteratively for resources.
  For example, returning MedicationRequest ‘mr123’, and the Patient resources it references and those patient's Practitioner:
`/MedicationRequest?_include=MedicationRequest:patient&_include:iterate=Patient:general-practitioner&_id=mr123`

The solution is based mainly on updating SqlQueryGenerator to store a dictionary of result resource type and the cte names associated with them. This way, when iterating over results, the selection is done from the relevant cte (or ctes in case of multiple results sets).

For this to work, the IncludeRewriter should have been updated to sort the search parameters so that an _include/_revinclude parameter will appear after all the expressions yielding the result sets it selects from. 

The feature doesn't fully implement the spec and some capabilities were left out following a discussion with the PG. 

**In scope of this PR:**
- **Supporting _include:iterate expressions, with single/multiple iterations and with wildcards.** 
   For example:
   - `/MedicationRequest?_include=MedicationRequest:patient&_include:iterate=Patient:general-practitioner&_id=mr123`
   - `MedicationRequest?_include:MedicationRequest:*&_include:iterate=Patient:general-practitioner `
   -  `MedicationRequest&_include:MedicationRequest:patient&_include:iterate=Patient:* MedicationRequest?_include:MedicationRequest:*&_include:iterate=Patient:*`

- **Supporting _revinclude:iterate expressions with single/multiple iterations and with wildcard.** 
  -  This is similar to the _include queries. However, for _revinclude we require that if the TargetType can be of multiple types, a resource type must be specified. Other a BadRequest is returned. For example:
  - ` Practitioner?_revinclude=Patient:*&_revinclude:iterate=MedicationRequest:patient`
  -  ` Practitioner?_revinclude:iterate=MedicationRequest:*&_revinclude=Patient:general-practitioner `
  -  `Patient?_revinclude:iterate=MedicationDispense:*&_revinclude=MedicationRequest:*`

**Out of scope of this PR:**
* #1310  
  Recursive iterate expressions are expression where the Source and the Target are of the same type. For example:  
  `Organization?_include:iterate=Organization:partof Patient&_include=Patient:organization&_include:iterate=Organization:partof `

* Supporting _revinclude:iterate expressions where there is more than a single target type and the target type is not specified:   
   Unsupported: `Patient?_revinclude=MedicationRequest:patient&_revinclude:iterate=MedicationDispense:performer`  
   Supported: `Patient?_revinclude=MedicationRequest:patient&_revinclude:iterate=MedicationDispense:performer:patient ` 
   Supported: `Patient?_revinclude=MedicationRequest:patient&_revinclude:iterate=MedicationDispense:prescription (since prescription can be only of type MedicationRequest) `

* _revinclude:iterate with WildCard. If the wildcard is associated with the _revinclude:iterate parameter, it is not supported. However, _revinclude:iterate can be used to iteratively query a results set of a wildcard expression:
    Unsupported: `Practitioner?_revinclude=Patient:general-practitioner&_revinclude:iterate=MedicationRequest:* `
   Supported: Practitioner?_revinclude=Patient:*&_revinclude:iterate=MedicationRequest:patient 

* #1309 
  The current implementation limits the item count of each included result set. In case where _include:iterate (or _revinclude:iterate) is applied on more than one result set, each application is limited.
For example: `MedicationDispense?_include=MedicationDispense:patient&_include=MedicationDispense:prescription&_include:iterate=MedicationRequest:patient&_include:iterate=Patient:general-practitioner`
Here, '_include:iterate=Patient:general-practitioner' refers to two included result sets (ctes): `MedicationDispense:patient' and `MedicationRequest:patient`. So each selection will be limited. 

## Related issues
Addresses #1313

## Testing
* IncludeRewriter unit tests to ensure the expressions are ordered correctly before executing them 
* IncludeSearch E2E tests are added for all the supported cases
* IncludeSearch E2E test case is added to ensure a BadRequest is returned in case of _revinclude:iterate expression is used with multiple target resource types without specifying one.

limorl commented:
FYI @eladiw

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -453,10 +517,59 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                     }
 
+                    // Update target reference cte dictionary
+                    // var curCte = TableExpressionName(_tableExpressionCounter);
```
redundant code

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -381,6 +389,10 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                 case TableExpressionKind.Include:
                     var includeExpression = (IncludeExpression)tableExpression.NormalizedPredicate;
 
+                    _includeCtes = _includeCtes ?? new List<string>();
```
This one is actually a list of ids, same for the _includeFromCtes.
can we have this in the var name to make it easier to read

limorl replied:
I'll rename to _includeCteNames - I think Name is better than Id in case of cte

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                                    fromCte = _includeFromCtes[++_curFromCteIndex];
+                                }
+                            }
+
```
empty line

limorl replied:
No problem with that :)

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                                }
+                                else if (includeExpression.ReferenceSearchParameter?.TargetResourceTypes != null)
+                                {
+                                    // Assumes TargetResourceTypes is of length 1. Otherwise, a BadRequest would have been thrown earlier for _revinclude:iterate
```
I would also throw here in any case, in the future someone might remove the other exception throw without knowing about this dependency.

limorl replied:
No need to, that's why we have a test for it :)

If someone removes it in SearchOpetionsFactory, the test will fail :)

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -453,10 +517,59 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                    {
+                        StringBuilder.Append($"),{Environment.NewLine}");
+
+                        // If it's not the last result set, append a new IncludeLimit cte, since IncludeLimitCte was not created for the current cte
```
why if its the last you dont add another limit? because anyway at the end we add one?

limorl replied:
Because an IncludeLimit expression is added after every IncludeExpression in IncludeReqriter

eladiw commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -453,10 +517,59 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                 case TableExpressionKind.IncludeLimit:
                     StringBuilder.Append("SELECT DISTINCT ");
 
+                    // TODO - Limiting the results of both _include and _revinclude parameters even though _revinclude is more risky
```
issue:
https://github.com/microsoft/fhir-server/issues/1309

eladiw reviewed (changes requested):
Great work.
I had some questions and comments.

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer.UnitTests/Microsoft.Health.Fhir.SqlServer.UnitTests.csproj:
```diff
@@ -9,5 +9,7 @@
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.Health.Fhir.SqlServer\Microsoft.Health.Fhir.SqlServer.csproj" />
+    <ProjectReference Include="..\Microsoft.Health.Fhir.Core\Microsoft.Health.Fhir.Core.csproj" />
+    <ProjectReference Include="..\Microsoft.Health.Fhir.Stu3.Core\Microsoft.Health.Fhir.Stu3.Core.csproj" />
```
Is it possible to avoid the version specific reference here?

limorl replied:
That was the only way I could get it work. Do you have any suggestion? perhaps I haven't tried it

limorl replied:
See my reply above,

limorl commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -273,6 +284,34 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
 
             return searchOptions;
 
+            IEnumerable<IncludeExpression> ParseIncludeIterateExpressions(SearchParams searchParams, bool reversed)
```
See my previous answer

eladiw replied:
but why is the nesting needed method declared within another method

limorl replied:
Not my choice, just followed the same pattern used with             void ValidateTotalType(TotalType totalType).
I can un-nest both of them.

Perhaps @brendankowitz knows why it was nested in the first place

limorl commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -191,17 +194,25 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
+            // See https://github.com/FirelyTeam/fhir-net-api/issues/222
+            // _include:iterate (_include:recurse) expression may appear without a preceding _include parameter
+            // when applied on a circular reference
+            searchExpressions.AddRange(ParseIncludeIterateExpressions(searchParams, false));
```
It basically checks the SearchParams.Parameters for _include:iterate parameters and then for _revinclude:iterate parameters.

I'll unify them into one call searchExpressions.AddRange(ParseIncludeIterateExpressions(searchParams) and check inside for _include and _revinclude parameters.

limorl commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -273,6 +284,34 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
+                .Where(p => p != null && iterateModifiers.Where(m => string.Equals(p.Item1, m, StringComparison.OrdinalIgnoreCase)).Any())
+                .Select(p =>
+                {
+                    ResourceType parsedIncludeResourceType = ResourceType.DomainResource;
```
No, it will not compile without it because of this if statement:

`if (!string.IsNullOrWhiteSpace(includeResourceType) &&
     !Enum.TryParse(includeResourceType, out parsedIncludeResourceType))
     {
            throw new ResourceNotSupportedException(includeResourceType);
      }

limorl commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -27,40 +28,89 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
             }
 
-            bool containsInclude = false;
+            List<TableExpression> reorderedExpressions = expression.TableExpressions.OrderByDescending(t => t, new TableExpressionComparer()).ToList();
```
Good catch!

limorl commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTestFixture.cs:
```diff
@@ -91,27 +115,156 @@ Observation CreateObservation(Patient patient, CodeableConcept code)
+                                Actor = new ResourceReference($"Practitioner/{Practitioner.Id}"),
+                            },
+                        },
+#if R5
```
I see your point, but was trying to avoid maintaining these tests in 3 different projects and duplicating the Fixture as well. Do you still think it's better?

limorl replied:
I didn’t find a way to do it without duplicating all the common tests.
Not sure it’s better, unless I am missing something

limorl commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+
+                                    if (TryGetIncludeCtes(targetType, out fromCtes))
+                                    {
+                                            _includeFromCtes.AddRange(fromCtes);
```
Fixed indentation but didn't see an issue with empty lines (also, it wouldn't compile if there were)

limorl commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -566,6 +679,27 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+        {
+            StringBuilder.Append(TableExpressionName(++_tableExpressionCounter)).AppendLine(" AS").AppendLine("(");
+
+            // the related cte is a reverse include, limit the number of returned items and count to
```
It does destinguish. It's within if(isRev){..}

There's an issue for limiting also _include and not only _revinclude: #1309

limorl commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -628,6 +762,33 @@ private void AppendHistoryClause(in IndentedStringBuilder.DelimitedScope delimit
             }
         }
 
+        private void AddIncludeLimitCte(string resourceType, string cte)
```
See your point. I used 'Append' to indicate StringBuilder.Append.
 I'll change it to *WriteIncludeLimitCte*

eladiw commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -191,17 +194,25 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
             {
                 searchExpressions.AddRange(searchParams.Include.Select(
-                    q => _expressionParser.ParseInclude(parsedResourceType.ToString(), q, false /* not reversed */))
+                    q => _expressionParser.ParseInclude(parsedResourceType.ToString(), q, false /* not reversed */, false))
```
sounds better.

eladiw commented on src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/ExpressionParser.cs:
```diff
@@ -101,7 +102,14 @@ public IncludeExpression ParseInclude(string resourceType, string includeValue,
             }
 
-            return new IncludeExpression(resourceType, refSearchParameter, originalType.ToString(), targetType, wildCard, isReversed);
+            referencedTypes = new List<string>();
+            _searchParameterDefinitionManager.GetSearchParameters(resourceType)
+            .Where(p => p.Type == ValueSets.SearchParamType.Reference).ToList()
+            .ForEach(p => referencedTypes.AddRange(p.TargetResourceTypes?.ToList()));
+
+            referencedTypes = referencedTypes.Distinct().ToList();
```
what about this, should this be done only if iterate == true?

limorl replied:
De facto, it is used for expressions that are wildcards. And then :iterate expressions that come after can know what are the possible ctes to select from.

eladiw reviewed (approved):
LGTM

johnstairs commented on src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/ExpressionParser.cs:
```diff
@@ -101,7 +102,14 @@ public IncludeExpression ParseInclude(string resourceType, string includeValue,
-            return new IncludeExpression(resourceType, refSearchParameter, originalType.ToString(), targetType, wildCard, isReversed);
+            referencedTypes = new List<string>();
+            _searchParameterDefinitionManager.GetSearchParameters(resourceType)
+            .Where(p => p.Type == ValueSets.SearchParamType.Reference).ToList()
```
This looks pretty allocation-heavy. Also, it seems like something we could cache (and populate lazily).

limorl replied:
Agree. The challenge was to propagate the searchParameterDefinitionManager to the IncludeExpression, so that it will be used upon need. That didn't make much sense to me since IncludeExpression represents a parsed expression, so ideally searchParameterDefinitionManager shouldn't be used in it. 

Maybe we can use a delegate that is applied once when ReferencedTypes is used and populates it. WDYT?

limorl replied:
Actually, after giving it another thought, It’s worth noting that it will always be used when the expression is _include wildcard followed by an /include iterate (or revinclude), since I am storing the cte names in the dictionary.

So I suggest to always populate it for wildcard followed by  iterate expression.
So basically the search options factory “knows” if there are include iterate expressions.

WDYT?

johnstairs commented on src/Microsoft.Health.Fhir.Core/Resources.Designer.cs:
```diff
@@ -132,9 +132,6 @@ internal class Resources {
             }
         }
         
```
Was this removed by mistake?

limorl replied:
Yes, my bad. Fixed now.

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                                {
+                                    // Assumes TargetResourceTypes is of length 1. Otherwise, a BadRequest would have been thrown earlier for _revinclude:iterate
+                                    List<string> fromCtes;
+                                    var targetType = includeExpression.ReferenceSearchParameter.TargetResourceTypes.First();
```
Should we use `.Single()` to enforce this assumption?

limorl replied:
We could, but as I answered to Elad, if there are multiple target types and one is not specified, a BadRequest is thrown earlier and this is validated in an E2E test. We have E2E test to validate that.

Moreover, if you enforce it here (SQLQueryGenerator class), we will need to throw a BadRequest for throw without catching which will end up in an internal error. Both are not ideal IMO. We assume a valid request by the time we get to SqlQueryGenerator. Notice that we don;t  throw BadRequest from any code path in this class.

WDYT?

brendankowitz replied:
Agree a Bad Request should be thrown before we get here. I'm happy to keep First(), but perhaps lets Log or Assert something if the expectation is not met? If things change upstream in future somehow, this has the potential to return the wrong information

limorl replied:
If things change in the future, a test will fail (I added it myself).
But I agree with you, logging is a good way to address it. 
The problem though is that there's no logger in SqlQueryGenerator class, so either one of the two:
1. Add logger to SqlQueryGenerator class (dependency injection) 
2. Through and catch at the SqlServerSearchService  and log 

I don;t think any of the option is good.
WDYT?

brendankowitz replied:
If we have a test to validate, happy to leave it with this for now

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -453,10 +517,58 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                    // Handle Multiple Results sets to include from
+                    if (_includeFromCteIds?.Count > 1 && _curFromCteIndex >= 0 && _curFromCteIndex < _includeFromCteIds.Count - 1)
+                    {
+                        StringBuilder.Append($"),{Environment.NewLine}");
```
Is there an `AppendLine`?

limorl replied:
There is. I wanted to be consistent with existing code earlier in that class:
                StringBuilder.AppendDelimited($",{Environment.NewLine}", expression.TableExpressions, (sb, tableExpression) =>

johnstairs commented:
It looks like the link url in the bundle is leaving out the `:iterate` parameters:

```http
GET https://localhost:44348/MedicationRequest?_include=MedicationRequest:patient&_include:iterate=Patient:general-practitioner&_id=123
```

``` JSON
{
  "resourceType": "Bundle",
  "id": "949805aa-9bc2-4195-b581-98df15005424",
  "meta": {
    "lastUpdated": "2020-10-09T20:29:01.8167293+00:00"
  },
  "type": "searchset",
  "link": [
    {
      "relation": "self",
      "url": "https://localhost:44348/MedicationRequest?_include=MedicationRequest%3Apatient&_id=123"
    }
  ]
}
```

johnstairs commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -273,6 +283,35 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
 
+            IEnumerable<IncludeExpression> ParseIncludeIterateExpressions(SearchParams searchParams)
+            {
+                var allIterateModifiers = IncludeIterateModifiers.Concat(RevIncludeIterateModifiers).ToList();
```
There's no need to compute this every time

limorl replied:
Right! Fixed.

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -27,19 +28,17 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
+            // Order expressions so that simple search parameters appear first and include parameters appear after
+            // In addition, sort that _include:iterate and _revinclude:iterate parameters are ordered so that the
+            // included results they reference appears before
+            List<TableExpression> reorderedExpressions = expression.TableExpressions.OrderByDescending(t => t, new TableExpressionComparer()).ToList();
```
I realize it was like this before, but it's a tad confusing that the comparer reverses the return value and we are ordering by descending.

johnstairs replied:
Also please reuse a static instance of TableExpressionComparer

limorl replied:
I agree it's counter intuitive, I was trying to be consistent with prev code.
I can definitely remove the OrderByDescending and reverse the comparer accordingly.

limorl replied:
Both fixed.

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -27,19 +28,17 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
             }
 
-            bool containsInclude = false;
+            bool containsInclude = expression.TableExpressions.AsEnumerable().Where(e => e.SearchParameterQueryGenerator is IncludeQueryGenerator).Any();
```
What is the scenario when `Kind == Include` and  `!(SearchParameterQueryGenerator is IncludeQueryGenerator)`

limorl replied:
TBH I don't know that there is. I was consistent with the code that existed before. 
I'll try to use Kind and make sure all tests pass

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -27,19 +28,17 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
             }
 
-            bool containsInclude = false;
+            bool containsInclude = expression.TableExpressions.AsEnumerable().Where(e => e.SearchParameterQueryGenerator is IncludeQueryGenerator).Any();
```
BTW, expressions like this could be simplified to 
`.Any(e => e.SearchParameterQueryGenerator is IncludeQueryGenerator)`

limorl replied:
Cool! Thanks. Done.

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -55,12 +54,66 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
+
+                    // Both expressions are Include:iterate expressions, order so that _include:iterate source type will appear after relevant include target
+                    var xTargetTypes = !string.IsNullOrEmpty(xInclude.TargetResourceType) ? new List<string>() { xInclude.TargetResourceType } : xInclude.ReferenceSearchParameter.TargetResourceTypes;
+                    var yTargetTypes = !string.IsNullOrEmpty(yInclude.TargetResourceType) ? new List<string>() { yInclude.TargetResourceType } : yInclude.ReferenceSearchParameter.TargetResourceTypes;
```
Please avoid allocating inside a comparer method

limorl replied:
Is there a known performance issue specific to this case?
I'll rewrite it without allocating a new list, but note that it will add several if cases.

limorl commented on src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/IncludeExpression.cs:
```diff
@@ -83,13 +99,12 @@ public IncludeExpression(string resourceType, SearchParameterInfo referenceSearc
+            var reversed = Reversed ? "Reversed" : string.Empty;
+            var wildcard = WildCard ? "Wildcard" : string.Empty;
+            var paramName = ReferenceSearchParameter != null ? ReferenceSearchParameter.Name : string.Empty;
+            return $"({reversed} Include {iterate} {wildcard} {paramName}{targetType})";
```
Agree. fixed.

limorl commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -191,17 +194,24 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
 
+            // Parse _include:iterate (_include:recurse) parameters.
+            // :iterate (:recurse) modifiers are not supported by Hl7.Fhir.Rest, hence not added to the Include collection and exist in the Parameters list.
+            // See https://github.com/FirelyTeam/fhir-net-api/issues/222
```
Good news. I suggest to keep it as is untill v2 is used. Clearly moving to v2 should be out-of-scope for this PR

limorl commented on src/Microsoft.Health.Fhir.SqlServer.UnitTests/Features/Search/Expressions/IncludeRewriterTests.cs:
```diff
@@ -0,0 +1,225 @@
+            // Order the following query:
+            // [base]/MedicationDispense?_include:iterate=Patient:general-practitioner&_include:iterate=MedicationRequest:patient&_include=MedicationDispense:prescription&_id=smart-MedicationDispense-567
+
+            var refSearchParameter = _searchParameterDefinitionManager.GetSearchParameter("Patient", "general-practitioner");
```
The challenge is coming up with the URI.
For example, in Patient:general-practitioner, the URI is http://hl7.org/fhir/SearchParameter/Patient-general-practitioner.
If indeed it follows the same pattern for all the other resources and references then there shouldn't be a problem. 

I'll look into it.

brendankowitz replied:
I do agree in principle, however, by only linking to Stu3 it misses R4 and R5 Search parameter definitions which may be different. Keeping the DataLayer projects free of version specific references has been very helpful with maintaining multiple versions of FHIR. If we want the full coverage in all versions of FHIR would it make sense to move these to https://github.com/microsoft/fhir-server/tree/master/test/Microsoft.Health.Fhir.Shared.Tests.Integration? 
I know its a SQL Specific rewriter, but there isn't a SQL unit test project that offers this coverage, thoughts? //cc @johnstairs

limorl replied:
The way I see it, the R4 and R5 parameter definitions are not needed for this uni tests since we generated the relevant test cases for the unit tests using Stu3 parameters. We are not checking all the possible parameters and their version fit, we are checking a sorting algorithm... we need a few parameters (not all of them)

limorl commented:
> It looks like the link url in the bundle is leaving out the `:iterate` parameters:
> 

@johnstairs good catch! 
It's since the include iterate parameters appear in the unsupportedSearchParameters due to issue: https://github.com/FirelyTeam/fhir-net-api/issues/222.

I removed them from the unsupported parameters list and added validation tests.

limorl commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchOptionsFactory.cs:
```diff
@@ -273,6 +283,35 @@ public SearchOptions Create(string compartmentType, string compartmentId, string
+                    ResourceType parsedIncludeResourceType;
+                    var includeResourceType = p.Item2?.Split(':')[0];
+                    if (string.IsNullOrWhiteSpace(includeResourceType) ||
+                        !Enum.TryParse(includeResourceType, out parsedIncludeResourceType))
```
Cool. Thanks. Done.

johnstairs commented:
In `cte4` below, we apply a top clause on the resultset from `cte3` we use for the intersection:

``` SQL
      cte3 AS
      (
          SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
          FROM cte2
      ),
      cte4 AS
      (
          SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
          FROM dbo.ReferenceSearchParam refSource
          INNER JOIN dbo.Resource refTarget
          ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
              AND refSource.ReferenceResourceId = refTarget.ResourceId
          WHERE refSource.SearchParamId = @p5
              AND refTarget.IsHistory = 0
              AND refSource.IsHistory = 0
              AND refSource.ResourceTypeId = @p6
              AND refSource.ResourceSurrogateId IN (SELECT TOP(@p7) Sid1 FROM cte3)
      )
```

We should only be applying iterate to the first ten includes in `cte3`. (We've also discussed that we should apply an overall limit to included resources like we do for `_revinclude`, though I'm happy to take that as a separate PR)

johnstairs commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/IncludeSearchTests.cs:
```diff
@@ -444,13 +478,702 @@ public async Task GivenARevIncludeSearchExpressionWithNoReferences_WhenSearched_
+        public async Task GivenAnIncludeIterateSearchExpressionWithAdditionalParameters_WhenSearched_TheIterativeResultsShouldBeAddedToTheBundle()
+        {
+            // Non-recursive iteration - Single iteration (_include:iterate)
+            // _tag parameter is not used due to issue https://github.com/microsoft/fhir-server/issues/1235
```
This issue should be fixed now...

limorl replied:
BTW, the fix created another issue of getting invalid response for valid queries without _id parameter.
@eladiw  wrote to you about it

johnstairs replied:
Yeah, sorry about that.

johnstairs commented:
Parenthesis indentation is inconsistent (let's follow the style of `cte7`):
``` SQL
cte7 AS
(
    SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
    FROM cte6
),
cte8 AS
(
    SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
    FROM dbo.ReferenceSearchParam refSource
    INNER JOIN dbo.Resource refTarget
    ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
        AND refSource.ReferenceResourceId = refTarget.ResourceId
    WHERE refSource.SearchParamId = @p11
        AND refTarget.IsHistory = 0
        AND refSource.IsHistory = 0
        AND refSource.ResourceTypeId = @p12
        AND refSource.ResourceSurrogateId IN (SELECT TOP(@p13) Sid1 FROM cte3)
    ),
cte9 AS
    (
    SELECT DISTINCT TOP (@p14) Sid1, IsMatch, CASE WHEN count(*) over() > @p15 THEN 1 ELSE 0 END AS IsPartial
    FROM cte8
    ),
cte10 AS
    (
    SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
    FROM dbo.ReferenceSearchParam refSource
    INNER JOIN dbo.Resource refTarget
    ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
        AND refSource.ReferenceResourceId = refTarget.ResourceId
    WHERE refSource.SearchParamId = @p11
        AND refTarget.IsHistory = 0
        AND refSource.IsHistory = 0
        AND refSource.ResourceTypeId = @p12
        AND refSource.ResourceSurrogateId IN (SELECT TOP(@p16) Sid1 FROM cte7)
),
```

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -21,7 +22,13 @@ namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors.Q
     {
         private string _cteMainSelect; // This is represents the CTE that is the main selector for use with includes
-        private List<string> _includeCtes;
+        private List<string> _includeCteIds;
```
can this be a local variable?

limorl replied:
I don't see how.  It is used by all expression visits. .. Unless I don;t understand your question

limorl replied:
I don;t see how. It accumulates the ctes in each visit... unless I didn;t understand what you mean

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:
```diff
@@ -453,10 +517,58 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                    }
+
+                    // Handle Multiple Results sets to include from
+                    if (_includeFromCteIds?.Count > 1 && _curFromCteIndex >= 0 && _curFromCteIndex < _includeFromCteIds.Count - 1)
```
I feel like we could take complexity out of this class by figuring this stuff out in a visitor pass ahead of this. Each CTE we generate would be based on a TableExpression in the SqlRootExpression, and we could figure out, in the case of iterate expression, which TableExpression is is meant to build off of. Admittedly, taking direct references could be tricky as the trees are immutable and changing them requires replacing them. (We might want to use IDs and a symbol table).

limorl replied:
If I understand you correctly, then you mean building the dictionary of ctes to include from before all the visits, instead of building it on the fly. This is something I raised when we discussed recursive queries.

This will require a major  change in all the VisitSqlRoot method which I don;t think should be out of scope for this PR

johnstairs replied:
Yeah, let's leave it as it is now. I do feel like the overall complexity in this class is getting higher than I would like, so I'd like to do some refactoring in the future.

limorl commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -55,12 +50,116 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
+
+        private class TableExpressionComparer : IComparer<TableExpression>
+        {
+            public int Compare(TableExpression x, TableExpression y)
```
Yes, the purpose was to achieve topological sort of dependencies. 
What indirect dependency do you have in mind? I couldn't think of any.
Do you have an example of a query with indirect dependency?

Can we accept it in this PR and open an issue to implement a topological sort?

limorl replied:
Yes, I was trying to achieve topological sort. I see your point, but I don't see how we miss an indirect dependency with iterate. 
Say we have a < b < c search parameters. 

> In `cte4` below, we apply a top clause on the resultset from `cte3` we use for the intersection:
> 
> ```sql
>       cte3 AS
>       (
>           SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
>           FROM cte2
>       ),
>       cte4 AS
>       (
>           SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>           FROM dbo.ReferenceSearchParam refSource
>           INNER JOIN dbo.Resource refTarget
>           ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>               AND refSource.ReferenceResourceId = refTarget.ResourceId
>           WHERE refSource.SearchParamId = @p5
>               AND refTarget.IsHistory = 0
>               AND refSource.IsHistory = 0
>               AND refSource.ResourceTypeId = @p6
>               AND refSource.ResourceSurrogateId IN (SELECT TOP(@p7) Sid1 FROM cte3)
>       )
> ```
> 
> We should only be applying iterate to the first ten includes in `cte3`. (We've also discussed that we should apply an overall limit to included resources like we do for `_revinclude`, though I'm happy to take that as a separate PR)

True, but we agreed to have a separate issue for it: https://github.com/microsoft/fhir-server/issues/1309

johnstairs replied:
Regarding the TOP clause, I'm sorry my comment forgot a "not". We should not be applying the top clause here. We do not want to limit this to the first 10 includes.

limorl replied:
Right! Now it makes sense. I think we should limit it to the IncludeLimit though (default 100)

limorl replied:
As for topological sort, this will not work when we introduce a recursive queries (circular reference), but the comparator could actually work since we always put :iterate after _include/_revinclude. 

For example, this is a recursive query with circular reference since participant can be of type Practitioner. 
The comparator will actually help. 
CareTeam resource?
CareTeam?_include=CareTeam:participant&_include:iterate=Patient:general-practitioner

I'll try to see if I can create a test that fails on indirect dependency.
So for 3 parameters where a < b < c - I basically need to check all the possible queries (order permutations).
If all pass than I think we are safe with comparer

johnstairs replied:
We should maybe limit it, but then I suppose we should make sure we raise the "is partial" warning. Or we put no limit and let #1309 take care of limiting it.

limorl replied:
Looks like we are safe (for now).
Tried all 6 permutations of 3 parameters in this query and it was sorted correctly:
MedicationDispense?_include=MedicationDispense:prescription&_include:iterate=MedicationRequest:patient&_include:iterate=Patient:general-practitioner

johnstairs replied:
I think this is an example where this approach fails.
Correct results:
```
https://localhost:44348/MedicationDispense?_id=1&_include:iterate=Device:location&_include:iterate=Location:endpoint&_include=MedicationDispense:performer&_include:iterate=Patient:general-practitioner
```
Incorrect results (I just moved `_include:iterate=Device:location` to the back):
```
https://localhost:44348/MedicationDispense?_id=1&_include:iterate=Location:endpoint&_include=MedicationDispense:performer&_include:iterate=Patient:general-practitioner&_include:iterate=Device:location
```
[correct.txt](https://github.com/microsoft/fhir-server/files/5374716/correct.txt)
[incorrect.txt](https://github.com/microsoft/fhir-server/files/5374717/incorrect.txt)

johnstairs replied:
(BTW: I think it should be an error when you :iterate over a type that is not part the rest of the query)

limorl replied:
I think that's not a 'sort' issue, but as you said, it's due to having an "orphan" parameters that are not related to the query/

I investigated this case  when I started working on the feature, and this is how I found this issue: https://github.com/microsoft/fhir-server/issues/1267 
I left it out of scope since it is not an :iterate issue necessarily, but an  overall _include/_revinclude issue.

Hapi for example, doesn't throw an error, but rather ignores the "orphan" parts.

If we want to through an error, we could actually check for such "orphans" in the IncludeRewriter.

I don't think this should be handled in this specific PR.

Thoughts?

johnstairs replied:
Agree.

limorl replied:
Topological sort was implemented to order the parameters. 
Also, I have added the above queries to the unit tests. Note that all the parameters above are actually related to the query. Turns out MedicationDispense:performer includes Device.

limorl commented:
>
> Parenthesis indentation is inconsistent (let's follow the style of `cte7`):
> 

@johnstairs Yes, this is because SQLQueryGenerator uses  Microsoft.Health.SqlServer.IndentedStringBuilder which has no option to un-indent (removing chars from the end doesn;t work and is quite ugly).
I don't see how we can work around it.

Note that when calling recursively to AcceptVisitor, in the case where we are iterating over multiple results sets, that is done within an already indented expression sue to the main call:

               StringBuilder.**AppendDelimited(**$",{Environment.NewLine}", expression.TableExpressions, (sb, tableExpression) =>
                {
                    sb.Append(TableExpressionName(++_tableExpressionCounter)).AppendLine(" AS").AppendLine("(");

                    using (**sb.Indent()**)
                    {
                        tableExpression.AcceptVisitor(this, context);
                    }

                    sb.Append(")");
                });

limorl commented:
> Parenthesis indentation is inconsistent (let's follow the style of `cte7`):
> 
> ```sql
> cte7 AS
> (
>     SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
>     FROM cte6
> ),
> cte8 AS
> (
>     SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>     FROM dbo.ReferenceSearchParam refSource
>     INNER JOIN dbo.Resource refTarget
>     ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>         AND refSource.ReferenceResourceId = refTarget.ResourceId
>     WHERE refSource.SearchParamId = @p11
>         AND refTarget.IsHistory = 0
>         AND refSource.IsHistory = 0
>         AND refSource.ResourceTypeId = @p12
>         AND refSource.ResourceSurrogateId IN (SELECT TOP(@p13) Sid1 FROM cte3)
>     ),
> cte9 AS
>     (
>     SELECT DISTINCT TOP (@p14) Sid1, IsMatch, CASE WHEN count(*) over() > @p15 THEN 1 ELSE 0 END AS IsPartial
>     FROM cte8
>     ),
> cte10 AS
>     (
>     SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>     FROM dbo.ReferenceSearchParam refSource
>     INNER JOIN dbo.Resource refTarget
>     ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>         AND refSource.ReferenceResourceId = refTarget.ResourceId
>     WHERE refSource.SearchParamId = @p11
>         AND refTarget.IsHistory = 0
>         AND refSource.IsHistory = 0
>         AND refSource.ResourceTypeId = @p12
>         AND refSource.ResourceSurrogateId IN (SELECT TOP(@p16) Sid1 FROM cte7)
> ),
> ```

@johnstairs Yes, this is because SQLQueryGenerator uses  Microsoft.Health.SqlServer.IndentedStringBuilder which has no option to un-indent (removing chars from the end doesn;t work and is quite ugly).
I don't see how we can work around it.

Note that when calling recursively to AcceptVisitor, in the case where we are iterating over multiple results sets, that is done within an already indented expression sue to the main call:

               StringBuilder.AppendDelimited($",{Environment.NewLine}", expression.TableExpressions, (sb, tableExpression) =>
                {
                    sb.Append(TableExpressionName(++_tableExpressionCounter)).AppendLine(" AS").AppendLine("(");

                    using (sb.Indent())
                    {
                        tableExpression.AcceptVisitor(this, context);
                    }

                    sb.Append(")");
                });

limorl commented:
> In `cte4` below, we apply a top clause on the resultset from `cte3` we use for the intersection:
> 
> ```sql
>       cte3 AS
>       (
>           SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
>           FROM cte2
>       ),
>       cte4 AS
>       (
>           SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>           FROM dbo.ReferenceSearchParam refSource
>           INNER JOIN dbo.Resource refTarget
>           ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>               AND refSource.ReferenceResourceId = refTarget.ResourceId
>           WHERE refSource.SearchParamId = @p5
>               AND refTarget.IsHistory = 0
>               AND refSource.IsHistory = 0
>               AND refSource.ResourceTypeId = @p6
>               AND refSource.ResourceSurrogateId IN (SELECT TOP(@p7) Sid1 FROM cte3)
>       )
> ```
> 
> We should not only be applying iterate to the first ten includes in `cte3`. (We've also discussed that we should apply an overall limit to included resources like we do for `_revinclude`, though I'm happy to take that as a separate PR)

True, but we agreed it will be in a separated PR. We have an open issue for it: https://github.com/microsoft/fhir-server/issues/1309

johnstairs commented:
Regarding the indentation, let's revisit later. We can expose the indentation level as a property on the writer, but I think I would look at an approach that does not require recursion. But later...

johnstairs commented:
It looks like we still are applying a TOP clause on the intersection for non-iterate includes:
```SQL
      DECLARE @p0 SmallInt = 77;
      DECLARE @p1 VarChar(64) = '1';
      DECLARE @p2 Int = 11;
      DECLARE @p3 SmallInt = 828;
      DECLARE @p4 Int = 10;

      SET STATISTICS IO ON;
      SET STATISTICS TIME ON;

      WITH cte0 AS
      (
          SELECT ResourceSurrogateId AS Sid1
          FROM dbo.Resource
          WHERE IsHistory = 0
              AND IsDeleted = 0
              AND ResourceTypeId = @p0
              AND ResourceId = @p1

      ),
      cte1 AS
      (
          SELECT DISTINCT TOP (@p2) Sid1, 1 AS IsMatch, 0 AS IsPartial
          FROM cte0
          ORDER BY  Sid1 ASC
      ),
      cte2 AS
      (
          SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
          FROM dbo.ReferenceSearchParam refSource
          INNER JOIN dbo.Resource refTarget
          ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
              AND refSource.ReferenceResourceId = refTarget.ResourceId
          WHERE refSource.SearchParamId = @p3
              AND refTarget.IsHistory = 0
              AND refSource.IsHistory = 0
              AND refSource.ResourceTypeId = @p0
              AND refSource.ResourceSurrogateId IN (SELECT TOP(@p4) Sid1 FROM cte1)
      ),
      cte3 AS
      (
          SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
          FROM cte2
      ),
...
```

limorl commented:
> It looks like we still are applying a TOP clause on the intersection for non-iterate includes:
> 
> ```sql
>       DECLARE @p0 SmallInt = 77;
>       DECLARE @p1 VarChar(64) = '1';
>       DECLARE @p2 Int = 11;
>       DECLARE @p3 SmallInt = 828;
>       DECLARE @p4 Int = 10;
> 
>       SET STATISTICS IO ON;
>       SET STATISTICS TIME ON;
> 
>       WITH cte0 AS
>       (
>           SELECT ResourceSurrogateId AS Sid1
>           FROM dbo.Resource
>           WHERE IsHistory = 0
>               AND IsDeleted = 0
>               AND ResourceTypeId = @p0
>               AND ResourceId = @p1
> 
>       ),
>       cte1 AS
>       (
>           SELECT DISTINCT TOP (@p2) Sid1, 1 AS IsMatch, 0 AS IsPartial
>           FROM cte0
>           ORDER BY  Sid1 ASC
>       ),
>       cte2 AS
>       (
>           SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>           FROM dbo.ReferenceSearchParam refSource
>           INNER JOIN dbo.Resource refTarget
>           ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>               AND refSource.ReferenceResourceId = refTarget.ResourceId
>           WHERE refSource.SearchParamId = @p3
>               AND refTarget.IsHistory = 0
>               AND refSource.IsHistory = 0
>               AND refSource.ResourceTypeId = @p0
>               AND refSource.ResourceSurrogateId IN (SELECT TOP(@p4) Sid1 FROM cte1)
>       ),
>       cte3 AS
>       (
>           SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
>           FROM cte2
>       ),
> ...
> ```
> 
> I don't know if that's new from this PR...

This wasn't introduced by this PR. I only removed the TOP for iterate ctes since it includes from an IncludeLimit cte. 
See below the SQL query on master branch for this query:
`MedicationDispense?_include=MedicationDispense:prescription

I think we should change the non-iterate Top in a separate PR. The include limit feature is not tested  AFAIK and there's a bit of a challenge to test it by providing a test configuration. It deserves attention.
I think it should be part of this issue: https://github.com/microsoft/fhir-server/issues/1309

limorl commented:
/azp run

limorl commented:
@johnstairs  The build seems to fail on issues not related to the PR changes, but rather environment setup.
For example, provisioning environment, while a previous environment exists:
![Uploading image.png…]()

LTA-Thinking commented:
@limorl Sorry, your PR environment got stuck in a bad state due to a provisioning issue we were having last week. I'm cleaning up the environment and will kick off a new build.

LTA-Thinking commented:
/azp run

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -55,12 +59,164 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
+        private class IncludeExpressionTopologicalSort
+        {
+            // Based on Khan's algorithm. See https://en.wikipedia.org/wiki/Topological_sorting.
+            // The search queries are acyclic.
```
How do we know that they are acyclic?

limorl replied:
Perhaps a better way to say it was that we are not supporting cyclic queries (cyclic queries are actually recursive queries). According to the Fhir Team, it's up to the server implementor to decide how deep.
We agreed that recursive queries are out of scope. 

If it is a cyclic query (recursive query), we can throw an error. It's really a product decision.

See the FHIR chat: https://chat.fhir.org/#narrow/stream/179166-implementers/topic/_include.3Arecurse.20(_include.3Aiterate).20ambiguity

johnstairs replied:
Yes, I think we should throw an error if we encounter a cycle.

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -55,12 +59,164 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
+        {
+            // Based on Khan's algorithm. See https://en.wikipedia.org/wiki/Topological_sorting.
+            // The search queries are acyclic.
+            public static IList<TableExpression> Sort(IEnumerable<TableExpression> includeExpressions)
```
The complexity of this implementation is pretty high. My thought would have been to think of it in this way:

Suppose you have a query like
```
/MedicationDispense?_include=MedicationDispense:context&_include:iterate=Encounter:appointment&_include:iterate=Encounter:based-on&_include:iterate=Appointment:patient&_include:iterate=ServiceRequest:specimen
```
I would model it in this way:

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiBhcHBvaW50bWVudFtfaW5jbHVkZTppdGVyYXRlPUVuY291bnRlcjphcHBvaW50bWVudDxici8-UmVxdWlyZXM6IEVuY291bnRlcjxici8-IFByb2R1Y2VzOiBBcHBvaW50bWVudF1cbiBiYXNlZC1vbltfaW5jbHVkZTppdGVyYXRlPUVuY291bnRlcjpiYXNlZC1vbjxici8-UmVxdWlyZXM6IEVuY291bnRlcjxici8-IFByb2R1Y2VzOiBTZXJ2aWNlUmVxdWVzdF1cbiBwYXRpZW50W19pbmNsdWRlOml0ZXJhdGU9QXBwb2ludG1lbnQ6cGF0aWVudDxici8-UmVxdWlyZXM6IEFwcG9pbnRtZW50PGJyLz4gUHJvZHVjZXM6IFBhdGllbnRdXG4gc3BlY2ltZW5bX2luY2x1ZGU6aXRlcmF0ZT1TZXJ2aWNlUmVxdWVzdDpzcGVjaW1lbjxici8-UmVxdWlyZXM6IFNlcnZpY2VSZXF1ZXN0PGJyLz4gUHJvZHVjZXM6IFNwZWNpbWVuXVxuXG5cbiBwYXRpZW50LS0-YXBwb2ludG1lbnRcbiBzcGVjaW1lbi0tPmJhc2VkLW9uXG5cbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0IiwidGhlbWVWYXJpYWJsZXMiOnsiYmFja2dyb3VuZCI6IndoaXRlIiwicHJpbWFyeUNvbG9yIjoiI0VDRUNGRiIsInNlY29uZGFyeUNvbG9yIjoiI2ZmZmZkZSIsInRlcnRpYXJ5Q29sb3IiOiJoc2woODAsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsInByaW1hcnlCb3JkZXJDb2xvciI6ImhzbCgyNDAsIDYwJSwgODYuMjc0NTA5ODAzOSUpIiwic2Vjb25kYXJ5Qm9yZGVyQ29sb3IiOiJoc2woNjAsIDYwJSwgODMuNTI5NDExNzY0NyUpIiwidGVydGlhcnlCb3JkZXJDb2xvciI6ImhzbCg4MCwgNjAlLCA4Ni4yNzQ1MDk4MDM5JSkiLCJwcmltYXJ5VGV4dENvbG9yIjoiIzEzMTMwMCIsInNlY29uZGFyeVRleHRDb2xvciI6IiMwMDAwMjEiLCJ0ZXJ0aWFyeVRleHRDb2xvciI6InJnYig5LjUwMDAwMDAwMDEsIDkuNTAwMDAwMDAwMSwgOS41MDAwMDAwMDAxKSIsImxpbmVDb2xvciI6IiMzMzMzMzMiLCJ0ZXh0Q29sb3IiOiIjMzMzIiwibWFpbkJrZyI6IiNFQ0VDRkYiLCJzZWNvbmRCa2ciOiIjZmZmZmRlIiwiYm9yZGVyMSI6IiM5MzcwREIiLCJib3JkZXIyIjoiI2FhYWEzMyIsImFycm93aGVhZENvbG9yIjoiIzMzMzMzMyIsImZvbnRGYW1pbHkiOiJcInRyZWJ1Y2hldCBtc1wiLCB2ZXJkYW5hLCBhcmlhbCIsImZvbnRTaXplIjoiMTZweCIsImxhYmVsQmFja2dyb3VuZCI6IiNlOGU4ZTgiLCJub2RlQmtnIjoiI0VDRUNGRiIsIm5vZGVCb3JkZXIiOiIjOTM3MERCIiwiY2x1c3RlckJrZyI6IiNmZmZmZGUiLCJjbHVzdGVyQm9yZGVyIjoiI2FhYWEzMyIsImRlZmF1bHRMaW5rQ29sb3IiOiIjMzMzMzMzIiwidGl0bGVDb2xvciI6IiMzMzMiLCJlZGdlTGFiZWxCYWNrZ3JvdW5kIjoiI2U4ZThlOCIsImFjdG9yQm9yZGVyIjoiaHNsKDI1OS42MjYxNjgyMjQzLCA1OS43NzY1MzYzMTI4JSwgODcuOTAxOTYwNzg0MyUpIiwiYWN0b3JCa2ciOiIjRUNFQ0ZGIiwiYWN0b3JUZXh0Q29sb3IiOiJibGFjayIsImFjdG9yTGluZUNvbG9yIjoiZ3JleSIsInNpZ25hbENvbG9yIjoiIzMzMyIsInNpZ25hbFRleHRDb2xvciI6IiMzMzMiLCJsYWJlbEJveEJrZ0NvbG9yIjoiI0VDRUNGRiIsImxhYmVsQm94Qm9yZGVyQ29sb3IiOiJoc2woMjU5LjYyNjE2ODIyNDMsIDU5Ljc3NjUzNjMxMjglLCA4Ny45MDE5NjA3ODQzJSkiLCJsYWJlbFRleHRDb2xvciI6ImJsYWNrIiwibG9vcFRleHRDb2xvciI6ImJsYWNrIiwibm90ZUJvcmRlckNvbG9yIjoiI2FhYWEzMyIsIm5vdGVCa2dDb2xvciI6IiNmZmY1YWQiLCJub3RlVGV4dENvbG9yIjoiYmxhY2siLCJhY3RpdmF0aW9uQm9yZGVyQ29sb3IiOiIjNjY2IiwiYWN0aXZhdGlvbkJrZ0NvbG9yIjoiI2Y0ZjRmNCIsInNlcXVlbmNlTnVtYmVyQ29sb3IiOiJ3aGl0ZSIsInNlY3Rpb25Ca2dDb2xvciI6InJnYmEoMTAyLCAxMDIsIDI1NSwgMC40OSkiLCJhbHRTZWN0aW9uQmtnQ29sb3IiOiJ3aGl0ZSIsInNlY3Rpb25Ca2dDb2xvcjIiOiIjZmZmNDAwIiwidGFza0JvcmRlckNvbG9yIjoiIzUzNGZiYyIsInRhc2tCa2dDb2xvciI6IiM4YTkwZGQiLCJ0YXNrVGV4dExpZ2h0Q29sb3IiOiJ3aGl0ZSIsInRhc2tUZXh0Q29sb3IiOiJ3aGl0ZSIsInRhc2tUZXh0RGFya0NvbG9yIjoiYmxhY2siLCJ0YXNrVGV4dE91dHNpZGVDb2xvciI6ImJsYWNrIiwidGFza1RleHRDbGlja2FibGVDb2xvciI6IiMwMDMxNjMiLCJhY3RpdmVUYXNrQm9yZGVyQ29sb3IiOiIjNTM0ZmJjIiwiYWN0aXZlVGFza0JrZ0NvbG9yIjoiI2JmYzdmZiIsImdyaWRDb2xvciI6ImxpZ2h0Z3JleSIsImRvbmVUYXNrQmtnQ29sb3IiOiJsaWdodGdyZXkiLCJkb25lVGFza0JvcmRlckNvbG9yIjoiZ3JleSIsImNyaXRCb3JkZXJDb2xvciI6IiNmZjg4ODgiLCJjcml0QmtnQ29sb3IiOiJyZWQiLCJ0b2RheUxpbmVDb2xvciI6InJlZCIsImxhYmVsQ29sb3IiOiJibGFjayIsImVycm9yQmtnQ29sb3IiOiIjNTUyMjIyIiwiZXJyb3JUZXh0Q29sb3IiOiIjNTUyMjIyIiwiY2xhc3NUZXh0IjoiIzEzMTMwMCIsImZpbGxUeXBlMCI6IiNFQ0VDRkYiLCJmaWxsVHlwZTEiOiIjZmZmZmRlIiwiZmlsbFR5cGUyIjoiaHNsKDMwNCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGUzIjoiaHNsKDEyNCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIiwiZmlsbFR5cGU0IjoiaHNsKDE3NiwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGU1IjoiaHNsKC00LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkiLCJmaWxsVHlwZTYiOiJoc2woOCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGU3IjoiaHNsKDE4OCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIn19LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbiBhcHBvaW50bWVudFtfaW5jbHVkZTppdGVyYXRlPUVuY291bnRlcjphcHBvaW50bWVudDxici8-UmVxdWlyZXM6IEVuY291bnRlcjxici8-IFByb2R1Y2VzOiBBcHBvaW50bWVudF1cbiBiYXNlZC1vbltfaW5jbHVkZTppdGVyYXRlPUVuY291bnRlcjpiYXNlZC1vbjxici8-UmVxdWlyZXM6IEVuY291bnRlcjxici8-IFByb2R1Y2VzOiBTZXJ2aWNlUmVxdWVzdF1cbiBwYXRpZW50W19pbmNsdWRlOml0ZXJhdGU9QXBwb2ludG1lbnQ6cGF0aWVudDxici8-UmVxdWlyZXM6IEFwcG9pbnRtZW50PGJyLz4gUHJvZHVjZXM6IFBhdGllbnRdXG4gc3BlY2ltZW5bX2luY2x1ZGU6aXRlcmF0ZT1TZXJ2aWNlUmVxdWVzdDpzcGVjaW1lbjxici8-UmVxdWlyZXM6IFNlcnZpY2VSZXF1ZXN0PGJyLz4gUHJvZHVjZXM6IFNwZWNpbWVuXVxuXG5cbiBwYXRpZW50LS0-YXBwb2ludG1lbnRcbiBzcGVjaW1lbi0tPmJhc2VkLW9uXG5cbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0IiwidGhlbWVWYXJpYWJsZXMiOnsiYmFja2dyb3VuZCI6IndoaXRlIiwicHJpbWFyeUNvbG9yIjoiI0VDRUNGRiIsInNlY29uZGFyeUNvbG9yIjoiI2ZmZmZkZSIsInRlcnRpYXJ5Q29sb3IiOiJoc2woODAsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsInByaW1hcnlCb3JkZXJDb2xvciI6ImhzbCgyNDAsIDYwJSwgODYuMjc0NTA5ODAzOSUpIiwic2Vjb25kYXJ5Qm9yZGVyQ29sb3IiOiJoc2woNjAsIDYwJSwgODMuNTI5NDExNzY0NyUpIiwidGVydGlhcnlCb3JkZXJDb2xvciI6ImhzbCg4MCwgNjAlLCA4Ni4yNzQ1MDk4MDM5JSkiLCJwcmltYXJ5VGV4dENvbG9yIjoiIzEzMTMwMCIsInNlY29uZGFyeVRleHRDb2xvciI6IiMwMDAwMjEiLCJ0ZXJ0aWFyeVRleHRDb2xvciI6InJnYig5LjUwMDAwMDAwMDEsIDkuNTAwMDAwMDAwMSwgOS41MDAwMDAwMDAxKSIsImxpbmVDb2xvciI6IiMzMzMzMzMiLCJ0ZXh0Q29sb3IiOiIjMzMzIiwibWFpbkJrZyI6IiNFQ0VDRkYiLCJzZWNvbmRCa2ciOiIjZmZmZmRlIiwiYm9yZGVyMSI6IiM5MzcwREIiLCJib3JkZXIyIjoiI2FhYWEzMyIsImFycm93aGVhZENvbG9yIjoiIzMzMzMzMyIsImZvbnRGYW1pbHkiOiJcInRyZWJ1Y2hldCBtc1wiLCB2ZXJkYW5hLCBhcmlhbCIsImZvbnRTaXplIjoiMTZweCIsImxhYmVsQmFja2dyb3VuZCI6IiNlOGU4ZTgiLCJub2RlQmtnIjoiI0VDRUNGRiIsIm5vZGVCb3JkZXIiOiIjOTM3MERCIiwiY2x1c3RlckJrZyI6IiNmZmZmZGUiLCJjbHVzdGVyQm9yZGVyIjoiI2FhYWEzMyIsImRlZmF1bHRMaW5rQ29sb3IiOiIjMzMzMzMzIiwidGl0bGVDb2xvciI6IiMzMzMiLCJlZGdlTGFiZWxCYWNrZ3JvdW5kIjoiI2U4ZThlOCIsImFjdG9yQm9yZGVyIjoiaHNsKDI1OS42MjYxNjgyMjQzLCA1OS43NzY1MzYzMTI4JSwgODcuOTAxOTYwNzg0MyUpIiwiYWN0b3JCa2ciOiIjRUNFQ0ZGIiwiYWN0b3JUZXh0Q29sb3IiOiJibGFjayIsImFjdG9yTGluZUNvbG9yIjoiZ3JleSIsInNpZ25hbENvbG9yIjoiIzMzMyIsInNpZ25hbFRleHRDb2xvciI6IiMzMzMiLCJsYWJlbEJveEJrZ0NvbG9yIjoiI0VDRUNGRiIsImxhYmVsQm94Qm9yZGVyQ29sb3IiOiJoc2woMjU5LjYyNjE2ODIyNDMsIDU5Ljc3NjUzNjMxMjglLCA4Ny45MDE5NjA3ODQzJSkiLCJsYWJlbFRleHRDb2xvciI6ImJsYWNrIiwibG9vcFRleHRDb2xvciI6ImJsYWNrIiwibm90ZUJvcmRlckNvbG9yIjoiI2FhYWEzMyIsIm5vdGVCa2dDb2xvciI6IiNmZmY1YWQiLCJub3RlVGV4dENvbG9yIjoiYmxhY2siLCJhY3RpdmF0aW9uQm9yZGVyQ29sb3IiOiIjNjY2IiwiYWN0aXZhdGlvbkJrZ0NvbG9yIjoiI2Y0ZjRmNCIsInNlcXVlbmNlTnVtYmVyQ29sb3IiOiJ3aGl0ZSIsInNlY3Rpb25Ca2dDb2xvciI6InJnYmEoMTAyLCAxMDIsIDI1NSwgMC40OSkiLCJhbHRTZWN0aW9uQmtnQ29sb3IiOiJ3aGl0ZSIsInNlY3Rpb25Ca2dDb2xvcjIiOiIjZmZmNDAwIiwidGFza0JvcmRlckNvbG9yIjoiIzUzNGZiYyIsInRhc2tCa2dDb2xvciI6IiM4YTkwZGQiLCJ0YXNrVGV4dExpZ2h0Q29sb3IiOiJ3aGl0ZSIsInRhc2tUZXh0Q29sb3IiOiJ3aGl0ZSIsInRhc2tUZXh0RGFya0NvbG9yIjoiYmxhY2siLCJ0YXNrVGV4dE91dHNpZGVDb2xvciI6ImJsYWNrIiwidGFza1RleHRDbGlja2FibGVDb2xvciI6IiMwMDMxNjMiLCJhY3RpdmVUYXNrQm9yZGVyQ29sb3IiOiIjNTM0ZmJjIiwiYWN0aXZlVGFza0JrZ0NvbG9yIjoiI2JmYzdmZiIsImdyaWRDb2xvciI6ImxpZ2h0Z3JleSIsImRvbmVUYXNrQmtnQ29sb3IiOiJsaWdodGdyZXkiLCJkb25lVGFza0JvcmRlckNvbG9yIjoiZ3JleSIsImNyaXRCb3JkZXJDb2xvciI6IiNmZjg4ODgiLCJjcml0QmtnQ29sb3IiOiJyZWQiLCJ0b2RheUxpbmVDb2xvciI6InJlZCIsImxhYmVsQ29sb3IiOiJibGFjayIsImVycm9yQmtnQ29sb3IiOiIjNTUyMjIyIiwiZXJyb3JUZXh0Q29sb3IiOiIjNTUyMjIyIiwiY2xhc3NUZXh0IjoiIzEzMTMwMCIsImZpbGxUeXBlMCI6IiNFQ0VDRkYiLCJmaWxsVHlwZTEiOiIjZmZmZmRlIiwiZmlsbFR5cGUyIjoiaHNsKDMwNCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGUzIjoiaHNsKDEyNCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIiwiZmlsbFR5cGU0IjoiaHNsKDE3NiwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGU1IjoiaHNsKC00LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkiLCJmaWxsVHlwZTYiOiJoc2woOCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwiZmlsbFR5cGU3IjoiaHNsKDE4OCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIn19LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

The edges arise from Requires to Produces, indicating a dependency. I would then do a DFS to create a topological sort. You would need a dictionary mapping a type name to the nodes that "produce it". 

What am I missing?

limorl replied:
That's pretty much what I did.  Only that I was looking at all combinations of  _include and _revinclude pairs.
Note that for _revinclude=MedicationDispense:prescription, the Requires and Produces are reversed.
It Requires MedicationRequest and Produces MedicationDispense.

So basically, the direction of the edge is determined by the IsDependencyEdge() function.
The runtime of my topological sort (Khan's algorithm) is O(V+E) and the implementation is clean and simple IMO.

limorl replied:
What we could do to make it cleaner and more intuitive is basically add IReadOnlyCollection<string> Requires and  IReadOnlyCollection<string>  Produces properties to IncludeExpression. This way IsDependencyEdge(x,y) is replaced with asking if x.Requires.Intersection(y.Produces).Any().

The sort itself is simple enough and of the same complexity of DFS O(V+E)

Are you ok with my suggestion above or should we keep as is?

johnstairs replied:
That is a lot clearer. I've expanded on it to show you what I had in mind: https://github.com/microsoft/fhir-server/commit/ca25c8b9eecb4c13fdf8e94df68811838b886a94. (Quickly implemented - would not claim it's properly tested. Also would need to add a proper error message for the cycle detection and a test)

limorl replied:
Nice commit! I like the a AssertAppearsBefore, I was thinking about implementing it, not sure why I haven't. 

I'll adopt some of your changes, but continue with Khan's implementation if that's ok. 

I will add the error.

johnstairs replied:
Go for it

johnstairs commented on src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs:
```diff
@@ -63,159 +63,103 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
-                        }
-                    }
-                }
+                throw new InvalidSearchOperationException("Cyclic include iterate queries are not allowed");
```
"not supported". And please put the string in a resx file.

limorl replied:
Oops.. Saw that after I merged. 
The reason I didn;t add the string to resx is that I wanted to be consistent with how all the Rewriters throw InvalidSearchOperationException.
